src/_posts/2019-10-30-crafting-websites-using-finite-state-machines-part-I.md

```diff
@@ -0,0 +1,403 @@
+                (init!))))
+```
+
+which will obviously
+
+Test fails! Good... now we write the code to make it pass:
+
+```clojure
+;THE MODEL
+(def model (atom {:context {:status "list"}}))
+
+
+;THE CONTROLLER
+(defn commit! [value atom-to-change]
+  (reset! atom-to-change value))
+
+(defn init! []
+  (-> @model
+    (assoc :todos {})
+    (assoc-in [:context :status] "list")
+    (commit! model)))
+
+
+
+;THE TEST
+(deftest initialize-in-list-mode
+         (is (= {:context {:status "list"} :todos {}}
+                (init!))))
+```
+
+[![TDD](https://img.youtube.com/vi/CvEoXuoAhCg/0.jpg)](https://www.youtube.com/watch?v=CvEoXuoAhCg)
+
+Any refactorings? No, let's move to the second test and so on until we solve the entire data problem, one test at a time, making sure it fails, then making it pass, then refactoring the code. We end up with this:
+
+![]({{site.baseurl}}/assets/custom/img/blog/2019-10-30-crafting-websites-using-finite-state-machines-part-I/crafting/controller%20tests.png "Controller tests")
+
+Solve the data problem (TDD the model/controller) Done!
+
+The code is merged here:
+https://gitlab.com/danbunea/how-to-solve-it-crafting-web-apps/merge_requests/1/diffs
+
+Code:
+
+- [The final code for the controller test](https://gitlab.com/danbunea/how-to-solve-it-crafting-web-apps/blob/master/test/todomvc/controller_should.cljs)
+- [The final code for the controller](https://gitlab.com/danbunea/how-to-solve-it-crafting-web-apps/blob/master/src/todomvc/controller.cljs)
+
+Now let's have another look again to see the correspondence between the transitions and the controller functions:
+
+![]({{site.baseurl}}/assets/custom/img/blog/2019-10-30-crafting-websites-using-finite-state-machines-part-I/crafting/controller%20to%20diagram.png "Controller to functions")
+
+while the data in the model corresponds to:
+
+![]({{site.baseurl}}/assets/custom/img/blog/2019-10-30-crafting-websites-using-finite-state-machines-part-I/crafting/model%20to%20diagram.png "Model to diagram")
+
+### Step 4: Build the UI
+
+We will need now to make sure that our states can be represented on the screen.
+
+For instance when we're in list mode, filtered:
+
+![]({{site.baseurl}}/assets/custom/img/blog/2019-10-30-crafting-websites-using-finite-state-machines-part-I/crafting/filtered%20list%20model%20to%20view.png "model to view, filtered mode")
+
+How about edit state:
+
+![]({{site.baseurl}}/assets/custom/img/blog/2019-10-30-crafting-websites-using-finite-state-machines-part-I/crafting/edit%20todo%20model%20to%20view.png "mode to view, edit mode")
+
+Now that we know how the data corresponds to the UI, we can move on to doing the actual UI. We'll use react/reagent to break the UI into components:
+
+![]({{site.baseurl}}/assets/custom/img/blog/2019-10-30-crafting-websites-using-finite-state-machines-part-I/crafting/component%20hierarchy%20and%20UI%20relations.png "component hierarchy")
+
+Then we'll test drive the entire UI starting by planning the tests:
```

Comment:
You are building first the UI and then adding the tests? That doesn't seem TDD.

Reply:
Very well spotted. The image is only obtained later, after writing the code TDD. But what I wanted to show is how everything is broken down into components, and usually I do that on paper. Maybe I need to specify this somehow

Reply:
Added a list of what is component hierarchy I am thinking of, and a hint that the image is obtained only after finishing the code.